toolchain/check/context.cpp

```diff
@@ -505,6 +532,102 @@ auto Context::ImplicitAsRequired(Parse::Node parse_node, SemIR::NodeId value_id,
   return output_value_id;
 }
 
+static auto CanConvertTupleToArray(Context& context, SemIR::Node tuple_type,
+                                   SemIR::Node array_type) -> bool {
+  auto [array_bound_id, element_type_id] = array_type.GetAsArrayType();
+  auto tuple_elem_types_id = tuple_type.GetAsTupleType();
+  const auto& tuple_elem_types =
+      context.semantics_ir().GetTypeBlock(tuple_elem_types_id);
+
+  if (tuple_elem_types.size() !=
+      context.semantics_ir().GetArrayBoundValue(array_bound_id)) {
+    return false;
+  }
+
+  // TODO: We should do a proper convertibility check here for each
+  // tuple element.
+  return llvm::all_of(tuple_elem_types,
+                      [element_type_id = element_type_id](auto type_id) {
+                        return type_id == element_type_id;
+                      });
+}
+
+// Performs a conversion from a tuple to an array type.
+static auto ConvertTupleToArray(Context& context, SemIR::Node tuple_type,
+                                SemIR::Node array_type,
+                                SemIR::TypeId array_type_id,
+                                SemIR::NodeId value_id) -> SemIR::NodeId {
+  auto [array_bound_id, element_type_id] = array_type.GetAsArrayType();
+  auto tuple_elem_types_id = tuple_type.GetAsTupleType();
+  const auto& tuple_elem_types =
+      context.semantics_ir().GetTypeBlock(tuple_elem_types_id);
+
+  auto array_bound = context.semantics_ir().GetArrayBoundValue(array_bound_id);
+  if (tuple_elem_types.size() != array_bound) {
+    CARBON_DIAGNOSTIC(
+        ArrayInitArgCountMismatch, Error,
+        "Cannot initialize array of {0} element(s) from {1} initializer(s).",
```

Comment:
Would this lead to:

```
var a: (i32,) = (0, 1);
var b: [i32; 1] = (1,);
// Cannot initialize array of 1 element(s) from 2 initializer(s).
b = a;
```

Does the "initialize" phrasing still align for you? (partly I comment because referring to the tuples as "N initializers" wasn't matching how I think of the code)

Reply:
That doesn't seem great; added a different diagnostic for the case where we're not initializing from a tuple literal.